examples/examples/tower_http.rs

```diff
@@ -0,0 +1,130 @@
+
+	fn on_request(&self, remote_addr: SocketAddr, request: &Request<Body>) -> Self::Instant {
+		println!("[Logger::on_request] remote_addr {}, request: {:?}", remote_addr, request);
+		Instant::now()
+	}
+
+	fn on_call(&self, name: &str, params: Params, kind: logger::MethodKind) {
+		println!("[Logger::on_call] method: '{}', params: {:?}, kind: {}", name, params, kind);
+	}
+
+	fn on_result(&self, name: &str, success: bool, started_at: Self::Instant) {
+		println!("[Logger::on_result] '{}', worked? {}, time elapsed {:?}", name, success, started_at.elapsed());
+	}
+
+	fn on_response(&self, result: &str, started_at: Self::Instant) {
+		println!("[Logger::on_response] result: {}, time elapsed {:?}", result, started_at.elapsed());
+	}
+}
+
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+	tracing_subscriber::FmtSubscriber::builder()
+		.with_env_filter(tracing_subscriber::EnvFilter::from_default_env())
+		.try_init()
+		.expect("setting default subscriber failed");
+
+	let addr = run_server().await?;
+	let url = format!("http://{}", addr);
+
+	let client = HttpClientBuilder::default().build(&url)?;
+	let response: String = client.request("say_hello", None).await?;
+	println!("[main]: response: {:?}", response);
+	let _response: Result<String, _> = client.request("unknown_method", None).await;
+	let _ = client.request::<String>("say_hello", None).await?;
+
+	Ok(())
+}
+
+async fn run_server() -> anyhow::Result<SocketAddr> {
+	// Construct a custom service for handling the RPC requests.
+	let make_service = make_service_fn(move |conn: &AddrStream| {
+		let remote_addr = conn.remote_addr();
+		async move {
+			let mut module = RpcModule::new(());
+			module.register_method("say_hello", |_, _| Ok("lo")).unwrap();
+
+			// Obtain the tower service relying on the RPC implementation.
+			// NOTE: RPC's logger can be chained with tower.
+			let rpc_svc = HttpServerBuilder::new().set_logger(Timings).to_service(module, remote_addr).unwrap();
+
+			// Chain multiple tower compatible layers on top of the RPC's service.
+			let tower_svc = tower::ServiceBuilder::new()
+				// Add high level tracing/logging to all requests
+				.layer(
+					TraceLayer::new_for_http()
+						.on_body_chunk(|chunk: &Bytes, latency: Duration, _: &tracing::Span| {
+							tracing::trace!(size_bytes = chunk.len(), latency = ?latency, "sending body chunk")
+						})
+						.make_span_with(DefaultMakeSpan::new().include_headers(true))
+						.on_response(DefaultOnResponse::new().include_headers(true).latency_unit(LatencyUnit::Micros)),
+				)
+				// Mark the `Authorization` request header as sensitive so it doesn't show in logs
+				.layer(SetSensitiveRequestHeadersLayer::new(once(hyper::header::AUTHORIZATION)))
+				.timeout(Duration::from_secs(2))
+				.service(rpc_svc);
+
+			Ok::<_, Infallible>(tower_svc)
+		}
+	});
+
+	let addr = SocketAddr::from(([127, 0, 0, 1], 9935));
+	tokio::spawn(async move { Server::bind(&addr).serve(make_service).await });
```

Comment:
this API is very low-level, can't we bind the addr and start the hyper server under the hood? 

I guess you wanted to avoid that with the `Option<SocketAddr>` but I think this is too low-level for must users.
Would be nice to just "plug-in" the tower service in the builder and then just call `server.start` IMO, is that hard or why did you end up with this approach?

Reply:
I think these trait bounds from [hyper 0.14.2](https://github.com/hyperium/hyper/blob/0ff6213be2c3b93a5cb0fa6caf169622131f8466/src/server/server.rs#L536) make it a lot harder than I expected to plug in the `tower::ServiceBuilder<_>` into the HttpBuilder or start method directly.

I agree this is a bit too low-level and takes a lot of tower knowledge to make it work.

I'll keep on trying to find a better way to pass the ServiceBuilder to our server 😄 

A direct passthrough version might look like the following, but still, there are some errors generated by `ConnStreamExec` from the above link

```rust
pub fn start_with_builder<T, U>(mut self, methods: impl Into<Methods>, builder: tower::ServiceBuilder<T>)
where
T: Layer<U>,
<T as Layer<U>>::Service: Send + Service<hyper::Request<Body>>,
<<T as Layer<U>>::Service as Service<hyper::Request<Body>>>::Error: Send + Sync + StdError,
```

Reply:
much nicer now :muscle: